01_oop_and_ood/src/Shape.ts

```diff
@@ -1,3 +1,72 @@
+import { Point } from './Point';
+
+export enum Color {
+  GREEN = 'green',
+  RED = 'red',
+}
+export enum FillStatus {
+  FILLED = 'filled',
+  NOT_FILLED = 'not filled',
+}
+export const AT_LEAST_3_POINTS_EXCEPTION =
+  'the `Shape` should have at least 3 points!';
+/**
+ * 2. Create abstract superclass called `Shape`, which contains:
+    - two protected instance variables: `color` (string), `filled`
+      (boolean) and `points` (Point[]);
+    - overloaded constructor (use multiple constructors declaration for Typescript): a
+      constructor that takes a list of `points` and initializes the color to `"green"`
+      and filled to `true` by default, and a constructor that takes a list of `points`,
+      the `color` and `filled` properties;
+    - Make sure that the `Shape` has at least 3 points (2 points is just a line).
+    - `toString()` method that returns `"A Shape with color of xxx and filled/Not
+     filled. Points: (x1, y1), (x2, y2)..."`;
+    - `getPerimeter()` that calculates the perimeter using `Point.distance` method;
+ */
 export abstract class Shape {
-    abstract getType(): string;
+  abstract getType(): string;
+
+  color: string;
+  filled: boolean;
+  points: Point[];
+
+  constructor(points: Point[]);
+  constructor(points: Point[], color?: string,  filled?: boolean) {
+    if (points.length < 3) {
+      throw AT_LEAST_3_POINTS_EXCEPTION;
```

Comment:
always better to wrap in new Error

```suggestion
      throw new Error(AT_LEAST_3_POINTS_EXCEPTION);
```